File: Sources/OpenVPN/CryptoOpenSSL_C/zeroing_data.c
//
//  zeroing_data.c
//  Partout
//
//  Created by Davide De Rosa on 6/14/25.
//  Copyright (c) 2025 Davide De Rosa. All rights reserved.
//
//  https://github.com/passepartoutvpn
//
//  This file is part of Partout.
//
//  Partout is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  Partout is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with Partout.  If not, see <http://www.gnu.org/licenses/>.
//

#include "allocation.h"
#include "zeroing_data.h"
#include <stdlib.h>
#include <string.h>

// MARK: Creation

zeroing_data_t *zd_create(size_t length) {
    zeroing_data_t *zd = pp_alloc_crypto(sizeof(zeroing_data_t));
    zd->bytes = pp_alloc_crypto(length);
    bzero(zd->bytes, length);
    zd->length = length;
    return zd;
}

zeroing_data_t *zd_create_copy(const uint8_t *bytes, size_t length) {
    zeroing_data_t *zd = pp_alloc_crypto(sizeof(zeroing_data_t));
    if (bytes) {
        zd->bytes = pp_alloc_crypto(length);
        memcpy(zd->bytes, bytes, length);
    }
    zd->length = length;
    return zd;
}

zeroing_data_t *zd_create_with_uint8(uint8_t value) {
    zeroing_data_t *zd = pp_alloc_crypto(sizeof(zeroing_data_t));
    zd->bytes = pp_alloc_crypto(1);
    zd->bytes[0] = value;
    zd->length = 1;
    return zd;
}

zeroing_data_t *zd_create_with_uint16(uint16_t value) {
    zeroing_data_t *zd = pp_alloc_crypto(sizeof(zeroing_data_t));
    zd->bytes = pp_alloc_crypto(2);
    zd->bytes[0] = value & 0xFF;
    zd->bytes[1] = value >> 8;
    zd->length = 2;
    return zd;
}

zeroing_data_t *zd_create_from_data(const uint8_t *data, size_t length) {
    if (!data || length == 0) return NULL;
    return zd_create_copy(data, length);
}

zeroing_data_t *zd_create_from_data_range(const uint8_t *data, size_t offset, size_t length) {
    if (!data || length == 0) return NULL;
    return zd_create_copy(data + offset, length);
}

zeroing_data_t *zd_create_from_string(const char *string, bool null_terminated) {
    if (!string) return NULL;
    size_t len = strlen(string);
    if (null_terminated) {
        return zd_create_copy((const uint8_t *)string, len + 1);
    } else {
        return zd_create_copy((const uint8_t *)string, len);
    }
}

void zd_free(zeroing_data_t *zd) {
    if (zd) {
        if (zd->bytes) {
            bzero(zd->bytes, zd->length);
            free(zd->bytes);
        }
        free(zd);
    }
}

// MARK: Properties

const uint8_t *zd_bytes(const zeroing_data_t *zd) {
    return zd ? zd->bytes : NULL;
}

uint8_t *zd_mutable_bytes(zeroing_data_t *zd) {
    return zd ? zd->bytes : NULL;
}

size_t zd_length(const zeroing_data_t *zd) {
    return zd ? zd->length : 0;
}

bool zd_equals(const zeroing_data_t *zd1, const zeroing_data_t *zd2) {
    if (!zd1 || !zd2) return false;
    return (zd1->length == zd2->length) && (memcmp(zd1->bytes, zd2->bytes, zd1->length) == 0);
}

// MARK: Copy

zeroing_data_t *zd_make_copy(const zeroing_data_t *zd) {
    if (!zd) return NULL;
    return zd_create_copy(zd->bytes, zd->length);
}

zeroing_data_t *zd_make_slice(const zeroing_data_t *zd, size_t offset, size_t length) {
    if (!zd || offset + length > zd->length) return NULL;

    zeroing_data_t *slice = pp_alloc_crypto(sizeof(zeroing_data_t));
    slice->bytes = pp_alloc_crypto(length);
    memcpy(slice->bytes, zd->bytes + offset, length);
    slice->length = length;
    return slice;
}

// MARK: Side effect

void zd_append(zeroing_data_t *zd, const zeroing_data_t *other) {
    if (!zd || !other) return;

    size_t new_len = zd->length + other->length;
    uint8_t *new_bytes = pp_alloc_crypto(new_len);
    memcpy(new_bytes, zd->bytes, zd->length);
    memcpy(new_bytes + zd->length, other->bytes, other->length);

    bzero(zd->bytes, zd->length);
    free(zd->bytes);
    zd->bytes = new_bytes;
    zd->length = new_len;
}

void zd_truncate(zeroing_data_t *zd, size_t new_length) {
    if (!zd || new_length > zd->length) return;

    uint8_t *new_bytes = pp_alloc_crypto(new_length);
    memcpy(new_bytes, zd->bytes, new_length);

    bzero(zd->bytes, zd->length);
    free(zd->bytes);

    zd->bytes = new_bytes;
    zd->length = new_length;
}

void zd_remove_until(zeroing_data_t *zd, size_t offset) {
    if (!zd || offset > zd->length) return;

    size_t new_length = zd->length - offset;
    uint8_t *new_bytes = pp_alloc_crypto(new_length);
    memcpy(new_bytes, zd->bytes + offset, new_length);

    bzero(zd->bytes, zd->length);
    free(zd->bytes);

    zd->bytes = new_bytes;
    zd->length = new_length;
}

void zd_zero(zeroing_data_t *zd) {
    if (zd && zd->bytes) {
        bzero(zd->bytes, zd->length);
    }
}

// MARK: Accessors

uint16_t zd_uint16(const zeroing_data_t *zd, size_t offset) {
    if (!zd || offset + 2 > zd->length) return 0;
    return zd->bytes[offset] | (zd->bytes[offset + 1] << 8);
}

File: Sources/OpenVPN/CryptoOpenSSL/Native/ZeroingData.swift
//
//  ZeroingData.swift
//  Partout
//
//  Created by Davide De Rosa on 6/14/25.
//  Copyright (c) 2025 Davide De Rosa. All rights reserved.
//
//  https://github.com/passepartoutvpn
//
//  This file is part of Partout.
//
//  Partout is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  Partout is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with Partout.  If not, see <http://www.gnu.org/licenses/>.
//

internal import _PartoutCryptoOpenSSL_C
import Foundation

public final class ZeroingData {
    private(set) var ptr: UnsafeMutablePointer<zeroing_data_t>?

    private init(ptr: UnsafeMutablePointer<zeroing_data_t>) {
        self.ptr = ptr
    }

    public init(length: Int = 0) {
        self.ptr = zd_create(length)
    }

    public init(bytes: UnsafePointer<UInt8>?, length: Int) {
        self.ptr = zd_create_from_data(bytes, length)
    }

    public init(uInt8: UInt8) {
        var value = uInt8
        self.ptr = zd_create_from_data(&value, 1)
    }

    public init(uInt16: UInt16) {
        var value = uInt16
        let ptr = withUnsafeBytes(of: &value) { $0.bindMemory(to: UInt8.self).baseAddress }
        self.ptr = zd_create_from_data(ptr, 2)
    }

    public init(data: Data) {
        data.withUnsafeBytes {
            self.ptr = zd_create_from_data($0.bindMemory(to: UInt8.self).baseAddress, data.count)
        }
    }

    public init(data: Data, offset: Int, length: Int) {
        data.withUnsafeBytes {
            self.ptr = zd_create_from_data_range($0.bindMemory(to: UInt8.self).baseAddress, offset, length)
        }
    }

    public init(string: String, nullTerminated: Bool) {
        guard let cstr = string.cString(using: .utf8) else {
            ptr = zd_create(0)
            return
        }
        self.ptr = zd_create_from_string(cstr, nullTerminated)
    }

    deinit {
        zd_free(ptr)
    }
}

// MARK: Properties

extension ZeroingData {
    public var bytes: UnsafePointer<UInt8>? {
        zd_bytes(ptr)
    }

    public var mutableBytes: UnsafeMutablePointer<UInt8>? {
        zd_mutable_bytes(ptr)
    }

    public var length: Int {
        Int(zd_length(ptr))
    }
}

extension ZeroingData: Equatable {
    public static func == (lhs: ZeroingData, rhs: ZeroingData) -> Bool {
        zd_equals(lhs.ptr, rhs.ptr)
    }
}

// MARK: Copy

extension ZeroingData {
    public func copy() -> ZeroingData {
        ZeroingData(ptr: zd_make_copy(self.ptr))
    }

    public func withOffset(_ offset: Int, length: Int) -> ZeroingData {
        ZeroingData(ptr: zd_make_slice(ptr, offset, length))
    }

    public func appending(_ other: ZeroingData) -> ZeroingData {
        guard let copy = zd_make_copy(ptr) else {
            return ZeroingData()
        }
        zd_append(copy, other.ptr)
        return ZeroingData(ptr: copy)
    }
}

// MARK: Side effect

extension ZeroingData {
    public func zero() {
        zd_zero(ptr)
    }

    public func truncate(toSize size: Int) {
        zd_truncate(ptr, size)
    }

    public func remove(untilOffset offset: Int) {
        zd_remove_until(ptr, offset)
    }

    public func append(_ other: ZeroingData) {
        zd_append(ptr, other.ptr)
    }
}

// MARK: Accessors

extension ZeroingData {
    public func networkUInt16Value(fromOffset offset: Int) -> UInt16 {
        CFSwapInt16BigToHost(zd_uint16(ptr, offset))
    }

    public func nullTerminatedString(fromOffset offset: Int) -> String? {
        precondition(offset <= length)
        guard let bytes else {
            return nil
        }
        return String(cString: bytes)
    }

    public func toData() -> Data {
        guard let ptr else {
            return Data()
        }
        return Data(bytes: ptr.pointee.bytes, count: ptr.pointee.length)
    }
}
